key-rotator/storage/manifest.go

```diff
@@ -0,0 +1,265 @@
+	key := m.keyFor(dataShareProcessorName)
+	manifestBytes, err := m.ds.get(ctx, key)
+	if err != nil {
+		return manifest.DataShareProcessorSpecificManifest{}, fmt.Errorf("couldn't get manifest from %q: %w", key, err)
+	}
+	var dspsm manifest.DataShareProcessorSpecificManifest
+	if err := json.Unmarshal(manifestBytes, &dspsm); err != nil {
+		return manifest.DataShareProcessorSpecificManifest{}, fmt.Errorf("couldn't unmarshal manifest from JSON: %w", err)
+	}
+	return dspsm, nil
+}
+
+// GetIngestorGlobalManifest gets the ingestor global manifest, if it exists
+// and is well-formed. If the manifest does not exist, an error wrapping
+// ErrObjectNotExist will be returned.
+func (m Manifest) GetIngestorGlobalManifest(ctx context.Context) (manifest.IngestorGlobalManifest, error) {
+	key := m.keyFor(ingestorGlobalManifestDataShareProcessorName)
+	manifestBytes, err := m.ds.get(ctx, key)
+	if err != nil {
+		return manifest.IngestorGlobalManifest{}, fmt.Errorf("couldn't get manifest from %q: %w", key, err)
+	}
+	var igm manifest.IngestorGlobalManifest
+	if err := json.Unmarshal(manifestBytes, &igm); err != nil {
+		return manifest.IngestorGlobalManifest{}, fmt.Errorf("couldn't unmarshal manifest from JSON: %w", err)
+	}
+	return igm, nil
+}
+
+func (m Manifest) keyFor(dataShareProcessorName string) string {
+	return path.Join(m.keyPrefix, fmt.Sprintf("%s-manifest.json", dataShareProcessorName))
+}
+
+// datastore represents a given key/value object store backing a Manifest. It
+// includes functionality for getting & putting individual objects by key,
+// specialized for small objects (i.e. no streaming support).
+type datastore interface {
+	// get gets the content of a given key, or returns an error if it can't.
+	// If the key does not exist, an error wrapping ErrObjectNotExist is
+	// returned.
+	get(ctx context.Context, key string) ([]byte, error)
+
+	// put puts the given content to the given key, or returns an error if it
+	// can't.
+	put(ctx context.Context, key string, data []byte) error
+}
+
+type gcsDatastore struct {
+	gcs    *storage.Client
+	bucket string
+}
+
+var _ datastore = gcsDatastore{} // verify gcsDatastore satsifies datastore.
+
+func (ds gcsDatastore) get(ctx context.Context, key string) ([]byte, error) {
+	r, err := ds.gcs.Bucket(ds.bucket).Object(key).NewReader(ctx)
+	if err != nil {
+		if err == storage.ErrObjectNotExist {
+			err = ErrObjectNotExist
+		}
+		return nil, fmt.Errorf("couldn't retrieve gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	defer r.Close() // errchecked Close is below
+	objBytes, err := io.ReadAll(r)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't read gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	if err := r.Close(); err != nil {
+		return nil, fmt.Errorf("couldn't close gs://%s/%s: %w", ds.bucket, key, err)
+	}
+	return objBytes, nil
+}
+
+func (ds gcsDatastore) put(ctx context.Context, key string, data []byte) error {
```

Comment:
We lost the structured log statement from `GCSStorage.writeManifest`

Reply:
Restored.